### ISSUE TITLE
Heap out of bounds read in cip_if_print()

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#567]
**Heap out of bounds read in cip_if_print()**

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r oobr_cip_if_print.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/20092218/30619766-a597-11e6-8a50-f6876fa1962d.png)

ASAN output:

```
==9996==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200000efb1 at pc 0x00000042aa41 bp 0x7fff33512310 sp 0x7fff33511aa0
READ of size 6 at 0x60200000efb1 thread T0
    #0 0x42aa40 in __interceptor_memcmp /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:629:7
    #1 0x5483df in cip_if_print XYZ/tcpdump/./print-cip.c:68:6
    #2 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #3 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #4 0x7adaf7 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #5 0x6b6f4c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #6 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #7 0x7fcbf1c9b82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #8 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

0x60200000efb1 is located 0 bytes to the right of 1-byte region [0x60200000efb0,0x60200000efb1)
allocated by thread T0 here:
    #0 0x4b880c in malloc /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64:3
    #1 0x7af103 in pcap_check_header XYZ/libpcap/./sf-pcap.c:401:14
    #2 0x7acf22 in pcap_fopen_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:400:7
    #3 0x7accaf in pcap_open_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:307:6
    #4 0x7fcbf1c9b82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:629:7 in __interceptor_memcmp
Shadow bytes around the buggy address:
  0x0c047fff9da0: fa fa 00 fa fa fa 00 fa fa fa 00 fa fa fa 07 fa
  0x0c047fff9db0: fa fa 00 fa fa fa 00 fa fa fa 05 fa fa fa 05 fa
  0x0c047fff9dc0: fa fa 07 fa fa fa 00 00 fa fa 00 00 fa fa 00 00
  0x0c047fff9dd0: fa fa 00 00 fa fa 00 00 fa fa 00 fa fa fa fd fd
  0x0c047fff9de0: fa fa fd fa fa fa fd fd fa fa fd fa fa fa fd fa
=>0x0c047fff9df0: fa fa fd fa fa fa[01]fa fa fa 00 07 fa fa fd fd
  0x0c047fff9e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==9996==ABORTING
```

**Comment by fxlb** [comment]
Thanks for the report.
This issue is already fixed in the security branch (Will be merged in a few weeks).

**Comment by infrastation** [comment]
The original reporter for this vulnerability is Hanno Böck. It took a long time to publish the fix so the same issue was later independently discovered by other researchers.

Commit 4ef024c fixes this and the fix is available in the release 4.9.0. Thanks to everybody for their reports, please send any future security reports to security@tcpdump.org.
